conans/model/dependencies.py

```diff
@@ -99,8 +99,12 @@ def from_node(node):
                 added_req.ref = RecipeReference.loads(old_req)
                 d[added_req] = existing
             # Now remove the replaced from dependencies dict
-            for new_req in node.replaced_requires.values():
-                d.pop(new_req, None)
+            for old_req, new_req in node.replaced_requires.items():
+                # But check we're not removing itself if old and new are the same ref
+                # Else we would delete the whole dependency
+                old_ref = RecipeReference.loads(old_req)
```

Comment:
Im actually not quite happy with this approach, but it's the best I could find - happy to hear alternatives!

Reply:
Did you check that `old_req` does not have `old_req.ref.name`?

Reply:
Yes, at this point old_req is a string key of the dictonary :)